src/main/kotlin/io/sip3/salto/ce/sdp/SdpHandler.kt

```diff
@@ -0,0 +1,174 @@
+    private val sessions = mutableMapOf<String, SdpSessionDescription>()
+
+    override fun start() {
+        config().getJsonObject("sip")?.getJsonObject("call")?.let { config ->
+            config.getLong("expiration-delay")?.let {
+                expirationDelay = it
+            }
+            config.getLong("duration-timeout")?.let {
+                durationTimeout = it
+            }
+        }
+
+        readCodecs(config())
+        vertx.eventBus().localConsumer<JsonObject>(RoutesCE.config_change) { event ->
+            try {
+                val config = event.body()
+                readCodecs(config)
+            } catch (e: Exception) {
+                logger.error("SdpHandler `readCodecs()` failed.", e)
+            }
+        }
+
+        vertx.setPeriodic(expirationDelay) {
+            terminateExpiredSessions()
+        }
+
+        vertx.eventBus().localConsumer<Pair<String, Any>>(RoutesCE.sdp_session) { event ->
+            try {
+                val (cmd, obj) = event.body()
+                when (cmd) {
+                    CMD_UPDATE -> handle(obj as SIPMessage)
+                    CMD_TERMINATE -> terminate(obj as String)
+                    else -> throw IllegalArgumentException("Unsupported SdpHandler command: '$cmd'")
+                }
+            } catch (e: Exception) {
+                logger.error("SdpHandler 'handle()' or 'terminate()' failed.", e)
+            }
+        }
+    }
+
+    private fun readCodecs(config: JsonObject) {
+        val tmpCodecs = mutableMapOf<String, Codec>()
+        config.getJsonArray("codecs")?.forEach { codecObject ->
+            val codec = (codecObject as JsonObject).mapTo(Codec::class.java)
+            tmpCodecs[codec.name] = codec
+        }
+        codecs = tmpCodecs
+    }
+
+    private fun terminateExpiredSessions() {
+        val now = System.currentTimeMillis()
+        sessions.filterValues { it.createdAt + durationTimeout < now }
+                .forEach { (transactionId, _) -> terminate(transactionId) }
+    }
+
+    private fun terminate(transactionId: String) {
+        sessions.remove(transactionId)
+    }
+
+    private fun handle(message: SIPMessage) {
+        val sessionDescription: SessionDescription = message.sessionDescription() ?: return
+
+        val transactionId = message.transactionId()
+        val session = sessions.getOrPut(transactionId) {
+            SdpSessionDescription().apply { this.callId = message.callId()!! }
+        }
+
+        when (message) {
+            is SIPRequest -> session.request = sessionDescription.getMediaDescription("audio")
+            is SIPResponse -> session.response = sessionDescription.getMediaDescription("audio")
+        }
+
+        if (session.request != null && session.response != null) {
+            session.codec = defineCodec(session.request!!, session.response!!)
+            send(session)
+        }
+    }
+
+    private fun send(session: SdpSessionDescription) {
```

Comment:
Let's better send it one by one. In this case we will save on creating this messy logic with listOf()